src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,289 @@
+
+        // Protected so tests can access
+        protected EditorState _currentState = EditorState.NoEditor;
+        private IVsWindowFrame _windowFrame;
+        private ITextBufferManager _textBufferManager;
+        private IProjectFileModelWatcher _projectFileModelWatcher;
+        private IFrameOpenCloseListener _frameEventsListener;
+        private ITextBufferStateListener _textBufferStateListener;
+
+        [ImportingConstructor]
+        public EditorStateModel(IProjectThreadingService threadingService,
+            UnconfiguredProject unconfiguredProject,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsShellUtilitiesHelper shellHelper,
+            ExportFactory<IProjectFileModelWatcher> projectFileModelWatcherFactory,
+            ExportFactory<ITextBufferStateListener> textBufferListenerFactory,
+            ExportFactory<IFrameOpenCloseListener> frameEventsListenerFactory,
+            ExportFactory<ITextBufferManager> textBufferManagerFactory)
+        {
+            Requires.NotNull(threadingService, nameof(threadingService));
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(shellHelper, nameof(shellHelper));
+            Requires.NotNull(projectFileModelWatcherFactory, nameof(projectFileModelWatcherFactory));
+            Requires.NotNull(textBufferListenerFactory, nameof(textBufferListenerFactory));
+            Requires.NotNull(frameEventsListenerFactory, nameof(frameEventsListenerFactory));
+            Requires.NotNull(textBufferManagerFactory, nameof(textBufferManagerFactory));
+
+            _threadingService = threadingService;
+            _unconfiguredProject = unconfiguredProject;
+            _serviceProvider = serviceProvider;
+            _shellHelper = shellHelper;
+            _projectFileWatcherFactory = projectFileModelWatcherFactory;
+            _textBufferListenerFactory = textBufferListenerFactory;
+            _frameEventsListenerFactory = frameEventsListenerFactory;
+            _textBufferManagerFactory = textBufferManagerFactory;
+        }
+
+        // Initialization Logic
+
+        /// <summary>
+        /// Called by anything attempting to open a project file editor window, usually by a command. This will show the window frame, creating it if
+        /// has not already been created.
+        /// </summary>
+        public async Task OpenEditorAsync()
+        {
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
+            {
+                // If the editor is already open, just show it and return
+                if (_currentState != EditorState.NoEditor)
+                {
+                    // If we're initializing, _windowFrame might be null. In that case, when the initialization code
+                    // is done, it'll take care of showing the frame.
```

Comment:
instead of that - why wouldn't we just check the _currentState up front? Why wouldn't we always let initializing show the window first?